homeassistant/components/fritzbox/config_flow.py

```diff
@@ -0,0 +1,143 @@
+"""Config flow for AVM Fritz!Box."""
+import socket
+from urllib.parse import urlparse
+
+from pyfritzhome import Fritzhome, LoginError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components.ssdp import ATTR_SSDP_LOCATION, ATTR_UPNP_FRIENDLY_NAME
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
+
+# pylint:disable=unused-import
+from .const import DEFAULT_HOST, DEFAULT_USERNAME, DOMAIN
+
+DATA_SCHEMA_USER = vol.Schema(
+    {
+        vol.Required(CONF_HOST, default=DEFAULT_HOST): str,
+        vol.Required(CONF_USERNAME, default=DEFAULT_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+    }
+)
+
+DATA_SCHEMA_CONFIRM = vol.Schema(
+    {
+        vol.Required(CONF_USERNAME, default=DEFAULT_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+    }
+)
+
+RESULT_AUTH_FAILED = "auth_failed"
+RESULT_NOT_FOUND = "not_found"
+RESULT_SUCCESS = "success"
+
+
+class FritzboxConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a AVM Fritz!Box config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+
+    def __init__(self):
+        """Initialize flow."""
+        self._host = None
+        self._manufacturer = None
+        self._model = None
+        self._name = None
+        self._password = None
+        self._username = None
+
+    def _get_entry(self):
+        return self.async_create_entry(
+            title=self._name,
+            data={
+                CONF_HOST: self._host,
+                CONF_PASSWORD: self._password,
+                CONF_USERNAME: self._username,
+            },
+        )
+
+    def _try_connect(self):
+        """Try to connect and check auth."""
+        fritzbox = Fritzhome(
+            host=self._host, user=self._username, password=self._password
+        )
+        try:
+            fritzbox.login()
+            fritzbox.logout()
+            return RESULT_SUCCESS
+        except OSError:
+            return RESULT_NOT_FOUND
+        except LoginError:
+            return RESULT_AUTH_FAILED
+
+    async def async_step_import(self, user_input=None):
+        """Handle configuration by yaml file."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+
+        if user_input is not None:
+            ip_address = await self.hass.async_add_executor_job(
+                socket.gethostbyname, user_input[CONF_HOST]
+            )
+
+            await self.async_set_unique_id(ip_address, raise_on_progress=False)
```

Comment:
We have decided that we shouldn't use the config entries unique_id api for ip addresses, since they are not completely stable.

Let's just implement a simple function that checks if there are already entries for this domain with the same address.

Reply:
Here's an example that does that and a bit more:
https://github.com/home-assistant/core/blob/7f06a939acefdcdf56d33cbb30ed7688edb31513/homeassistant/components/dynalite/config_flow.py#L25-L29

Reply:
This is reasonable but the result is the same. When the IP changes we are not able to detect if it's the same device.

Reply:
We only want to guard for duplicate entries. We don't want to use the unique_id api, since that implies that ids are both unique and stable.